test/doc/websocket_2_handshaking.cpp

```diff
@@ -132,6 +132,94 @@ snippets()
 
     //]
     }
+
+    {
+        //[code_websocket_2_6
+
+        // a function to select the most preferred protocol from a comma-separated list
+        auto best_protocol = [](string_view offered_tokens) -> std::string
+        {
+            // tokenize the Sec-Websocket-Protocol header offered by the client
+            http::token_list offered( offered_tokens );
+
+            // an array of protocols supported by this server
+            // in descending order of preference
+            static const std::array<string_view, 3>
+                supported = {
+                "v3.my.chat",
+                "v2.my.chat",
+                "v1.my.chat"
+            };
+
+            std::string result;
+
+            for (auto proto : supported)
+            {
+                auto iter = std::find(offered.begin(), offered.end(), proto);
+                if (iter != offered.end())
+                {
+                    // we found a supported protocol in the list offered by the client
+                    result.assign(proto.begin(), proto.end());
+                    break;
+                }
+            }
+
+            return result;
+        };
+
+
+        // This buffer is required for reading HTTP messages
+        flat_buffer buffer;
+
+        // Read the HTTP request ourselves
+        http::request<http::string_body> req;
+        http::read(sock, buffer, req);
+
+        // See if it's a WebSocket upgrade request
+        if(websocket::is_upgrade(req))
+        {
+            std::string selected_protocol =
+                best_protocol(
+                    req[http::field::sec_websocket_protocol]);
+
+            if (selected_protocol.empty())
+            {
+                // none of our supported protocols were offered
+                http::response<http::string_body> res;
+                res.result(http::status::bad_request);
+                res.body() = "No valid sub-protocol was offered."
+                              " This server implements"
+                              " v3.my.chat,"
+                              " v2.my.chat"
+                              " and v1.my.chat";
+                http::write(sock, res);
+            }
+            else
+            {
+                // Construct the stream, transferring ownership of the socket
+                stream<tcp_stream> ws(std::move(sock));
+
+                ws.set_option(
+                    stream_base::decorator(
+                        [selected_protocol](http::response_header<> &hdr) {
```

Comment:
This is rather annoying...A copy of `selected_protocol` exists for the lifetime of the `stream` (or until the decorator is changed). This isn't your problem but it does point out a design flaw.

Reply:
probably doesn't need to be a string in the sync version. But I'd rather have a string here otherwise I'm going to get a bunch of support emails from people using string_view objects in async initiators.